### PR TITLE
[Feature]: Add corrupted request metric to V1 metrics system.

### LINKED ISSUE #27301: [Feature]: Add num_corrupted_request metric to V1 metrics system.
### Description

Currently, vLLM internally tracks a corrupted_requests_counter metric whenever a request produces invalid outputs (NaNs) due to model, engine, or hardware issues. However, this metric is not directly exposed to users in logs or Prometheus metrics.

Exposing this metric would allow users to:
- Detect model instability or misbehaving custom models.
- Monitor runtime/engine health in production clusters.
- Quickly identify hardware or distributed inference issues affecting outputs

### Motivation & Problem

While NaN outputs are rare with well-tested models, they become critical for custom models in early development stages or may also arise within the runtime because of Engine/runtime issues. 

- Models may have numerical instability.
- Hardware issues are more likely to surface

The codebase already detects corrupted requests (Request.is_output_corrupted) when the env variable is set `VLLM_COMPUTE_NANS_IN_LOGITS = TRUE`, but this diagnostic information is completely hidden from users, as there are no metrics, no logging, and no monitoring.

### Proposed Idea

Two Approaches for Corrupted Request Metrics that i am thinking of are: 

1. Approach 1: CLI Config-Based (Current Implementation)

 -  `--include-corrupted-requests #CLI flag`
 - Config: SchedulerConfig.include_corrupted_requests
 - Usage: vllm serve model --include-corrupted-requests

Pros: User-friendly, explicit control, follows vLLM patterns
Cons: Adds new CLI argument, requires config changes

I welcome suggestions and thoughts on the same, and would love to contribute the same.

### Alternatives

2. Approach 2: Environment Variable-Based (Proposed Alternative)

- Existing `VLLM_COMPUTE_NANS_IN_LOGITS` environment variable
- Logic: When NaN detection is enabled, automatically expose corrupted metrics
- Usage: VLLM_COMPUTE_NANS_IN_LOGITS=1 vllm serve model

Pros: Reuses existing infrastructure, no new CLI args.
Cons: A coupling of metrics exposure to NaN detection, and less granular control.

Thanks
Snehlata

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by atalhens**
## 
Signed off: sneh.lata@nutanix.com
## Purpose

The purpose of this PR is to fix: https://github.com/vllm-project/vllm/issues/27301 and https://github.com/vllm-project/vllm/issues/27298. This PR implements a new Prometheus metric `vllm:num_requests_corrupted` to track and monitor corrupted requests in vLLM. The feature provides visibility into request corruption issues that can occur during inference, enabling better monitoring and debugging capabilities.

- New Prometheus gauge metric `vllm:num_requests_corrupted`
- Console logging of corrupted request counts per logging interval.
- Enabled via the env variable  `VLLM_COMPUTE_NANS_IN_LOGITS` flag for performance control, thus it won't impact any functionality unless enabled; the default value is set to False.
- Updated documentation and updated dashboard examples config as well.

**Performance Considerations:**
- Corrupted request counting is optional and disabled by default
- When enabled, it adds no computational overhead

## Test Plan
- Not needed as it is a by-default disabled metric.

## Test Result
- N/A

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: docs/design/metrics.md [hunk 1]
```diff
@@ -17,45 +17,39 @@ Metrics in vLLM can be categorized as follows:
...
-- `vllm:spec_decode_num_accepted_tokens_total` (Counter)
-- `vllm:spec_decode_num_draft_tokens_total` (Counter)
-- `vllm:spec_decode_num_emitted_tokens_total` (Counter)
+### v1 Metrics
+
+In v1, the following metrics are exposed via a Prometheus-compatible `/metrics` endpoint using the `vllm:` prefix:
+
+- `vllm:num_requests_running` (Gauge) - Number of requests currently running.
+- `vllm:num_requests_waiting` (Gauge) - Number of requests currently waiting.
+- `vllm:num_requests_corrupted` (Gauge) - Number of requests that became corrupted.
```
Comment (markmc) [inline docs/design/metrics.md]: Thanks, but I don't think it's super important for every new metric to be captured in this list - the design doc is a snapshot in time of a design discussion IMO

Let's complete the doc update in #27332

File: vllm/v1/metrics/loggers.py [hunk 2]
```diff
@@ -181,6 +181,7 @@ def log(self):
             "Avg generation throughput: %.1f tokens/s",
             "Running: %d reqs",
             "Waiting: %d reqs",
+            "Corrupted: %d reqs",
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py]: Let's only print this if `envs.VLLM_COMPUTE_NANS_IN_LOGITS`

File: vllm/v1/metrics/loggers.py [hunk 3]
```diff
@@ -376,6 +378,16 @@ def __init__(
             gauge_scheduler_waiting, engine_indexes, model_name
         )
 
+        gauge_scheduler_corrupted = self._gauge_cls(
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py]: Let's only register this if `envs.VLLM_COMPUTE_NANS_IN_LOGITS`

File: vllm/v1/core/sched/scheduler.py [hunk 4]
```diff
@@ -1154,9 +1155,22 @@ def update_draft_token_ids(
...
-    def get_request_counts(self) -> tuple[int, int]:
-        """Returns (num_running_reqs, num_waiting_reqs)."""
-        return len(self.running), len(self.waiting)
+    def _get_corrupted_requests_count(self, include_corrupted: bool) -> int:
+        """Get the count of corrupted requests if enabled, otherwise return 0.
+
+        This method centralizes the corrupted requests counting logic to avoid
+        code duplication and improve performance.
+        """
+        if not include_corrupted:
```
Comment (markmc) [inline vllm/v1/core/sched/scheduler.py]: Let's check `envs.VLLM_COMPUTE_NANS_IN_LOGITS` here

File: vllm/v1/core/sched/scheduler.py [hunk 5]
```diff
@@ -1154,9 +1155,22 @@ def update_draft_token_ids(
...
+            return 0
+        return sum(req.is_output_corrupted for req in self.running)
```
Comment (markmc) [inline vllm/v1/core/sched/scheduler.py]: I really don't like that we're doing this twice per step (AFAICT), but that's the fault of how `_maybe_publish_request_counts()` was implemented. IMHO, it should use the `SchedulerStats` that would have been prepared already

File: vllm/v1/metrics/loggers.py [hunk 6]
```diff
@@ -376,6 +378,16 @@ def __init__(
...
+            name="vllm:num_requests_corrupted",
+            documentation="Number of requests corrupted.",
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py]: It's more "Number of currently running requests that are corrupted" ?

File: vllm/engine/arg_utils.py [hunk 7]
```diff
@@ -1034,6 +1035,10 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
         scheduler_group.add_argument(
             "--async-scheduling", **scheduler_kwargs["async_scheduling"]
         )
+        scheduler_group.add_argument(
+            "--include-corrupted-requests",
```
Comment (markmc) [inline vllm/engine/arg_utils.py]: I don't think we need a command-line argument when we already have the env var

File: examples/online_serving/prometheus_grafana/grafana.json [hunk 8]
```diff
@@ -593,6 +593,22 @@
...
           "useBackend": false
+        },
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "${DS_PROMETHEUS}"
+          },
+          "disableTextWrap": false,
+          "editorMode": "builder",
+          "expr": "vllm:num_requests_corrupted{model_name=\"$model_name\"}",
```
Comment (markmc) [inline examples/online_serving/prometheus_grafana/grafana.json]: I don't think we should include such a niche, off-by-default, metric in the default dashboard

**Review by markmc — APPROVED** [review]
Thanks for all your work on this - it's pretty obvious that #18777 has very limited value without this

And ... I know the way `SchedulerStats.num_corrupted_reqs` is currently implemented as a "count of currently running requests that are corrupted" leads to using a `Gauge` ...

But I really think a `Counter` is a more natural use of a time-series metric for this information. Probably that would mean adding `is_output_corrupted` to `FinishedRequestStats` and removing `num_corrupted_reqs` from `SchedulerStats`?

**Review by markmc — CHANGES_REQUESTED** [review]
(sorry didn't mean to approve, wrong button!)

File: docs/design/metrics.md (hunk 1 above)
Comment (atalhens) [reply docs/design/metrics.md]: Hi Mark, thanks for the suggestions.
Surely I will skip over this change here, as a part of it has already been taken up in this https://github.com/vllm-project/vllm/pull/27332, and we can skip adding `vllm:num_requests_corrupted` in the list of items as a part of this PR.

FMI: 

> Let's complete the doc update in https://github.com/vllm-project/vllm/pull/27332

Should we consider adding this field in this PR, or are we thinking of skipping over it altogether? I do believe maybe not for every other metric, but this one seems important and has been in hindsight for a while due to https://github.com/vllm-project/vllm/pull/18777 not being exposed. Maybe adding it in there would give more visibility to the users.

File: vllm/engine/arg_utils.py (hunk 7 above)
Comment (atalhens) [reply vllm/engine/arg_utils.py]: > I don't think we need a command-line argument when we already have the env var

My thought behind adding was to tie it while doing vllm serve only to add ease, but it doesn't make sense, as it sounds better to use the env variable to avoid repetition. I also suggested this approach in the feature request issue. https://github.com/vllm-project/vllm/issues/27301.

Surely would refactor the code. Thanks for the suggestion

File: vllm/v1/metrics/loggers.py [hunk 9]
```diff
@@ -898,21 +916,12 @@ def record(
...
                 scheduler_stats.num_waiting_reqs
             )
-
             if self.show_hidden_metrics:
                 self.gauge_gpu_cache_usage[engine_idx].set(
                     scheduler_stats.kv_cache_usage
                 )
             self.gauge_kv_cache_usage[engine_idx].set(scheduler_stats.kv_cache_usage)
 
-            if self.show_hidden_metrics:
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py:943]: Removing these must be a mistake? Unrelated to the PR

File: vllm/v1/metrics/loggers.py [hunk 10]
```diff
@@ -103,6 +104,8 @@ def __init__(self, vllm_config: VllmConfig, engine_index: int = 0):
         self.last_generation_throughput: float = 0.0
         self.engine_is_idle = False
         self.aggregated = False
+        # Corrupted requests metric. This should not be reset.
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py]: I think it should be reset - we should only show the number of requests that were corrupted in the last logging interval?

**Review by markmc — CHANGES_REQUESTED** [review]

File: vllm/v1/metrics/loggers.py (hunk 9 above)
Comment (atalhens) [reply vllm/v1/metrics/loggers.py:943]: Ahh yes. Thanks for catching. Will revert

Comment (atalhens) [reply vllm/v1/metrics/loggers.py:943]: Removed the redundant check there.
`if self.show_hidden_metrics:`

Comment (markmc) [reply vllm/v1/metrics/loggers.py:943]: Please just revert any changes that are not required to add the corrupted request metric - there is no need to move the `kv_cache_usage` piece or remove the `show_hidden_metrics` check

File: vllm/v1/metrics/loggers.py [hunk 11]
```diff
@@ -903,6 +920,13 @@ def record(
                 self.gauge_gpu_cache_usage[engine_idx].set(
                     scheduler_stats.kv_cache_usage
                 )
+                self.counter_gpu_prefix_cache_queries[engine_idx].inc(
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py]: Not quite reverted yet

File: vllm/v1/core/sched/scheduler.py [hunk 12]
```diff
@@ -1261,7 +1262,6 @@ def make_stats(
             prefix_cache_stats=prefix_cache_stats,
             connector_prefix_cache_stats=connector_prefix_cache_stats,
             spec_decoding_stats=spec_decoding_stats,
-            num_corrupted_reqs=sum(req.is_output_corrupted for req in self.running),
```
Comment (markmc) [inline vllm/v1/core/sched/scheduler.py:1262]: The `is_output_corrupted()` helper can be removed

Comment (atalhens) [reply vllm/v1/core/sched/scheduler.py:1262]: IIUC, I have already removed the field altogether as it is not needed any longer. Could you please provide a little more detail in case you are pointing to something else?

Comment (markmc) [reply vllm/v1/core/sched/scheduler.py:1262]: This can be removed, because you have removed the only user of it:

```
class Request:
    @property
    def is_output_corrupted(self) -> bool:
        return self.num_nans_in_logits > 0
```

**Review by markmc — APPROVED** [review]
Thanks for your patience!

**Review by simon-mo — DISMISSED** [review]
@markmc reviewed

**Review by simon-mo — CHANGES_REQUESTED** [review]
Can you further justify the need for `VLLM_COMPUTE_NANS_IN_LOGITS` env var and why can't this be enabled by default? I cannot find a place where we were doing CPU <> GPU syncs in counting NaNs here.

**Comment by atalhens** [comment]
> > Can you further justify the need for `VLLM_COMPUTE_NANS_IN_LOGITS` env var and why can't this be enabled by default?
> 
> Looks like it was added in #18777 in response to [this comment](https://github.com/vllm-project/vllm/pull/18777#pullrequestreview-2941138025)
> 
> > I cannot find a place where we were doing CPU <> GPU syncs in counting NaNs here.
> 
> Yeah, looks like the original performance overhead concern wasn't very specific

Thanks for catching @simon-mo.
Correct, I think if we agree. I could remove the flag, which would make more sense.

**Review by simon-mo — APPROVED** [review]
Thank you for the benchmarks
